cmd/mmlogicapi/apisrv/apisrv.go

```diff
@@ -0,0 +1,659 @@
+
+	// Create context for tagging OpenCensus metrics.
+	funcName := "ListIgnoredPlayers"
+	fnCtx, _ := tag.New(c, tag.Insert(KeyMethod, funcName))
+
+	mlLog.WithFields(log.Fields{"ignorelist": ilName}).Info("Attempting to get ignorelist")
+
+	// retreive ignore list
+	il, err := ignorelist.Retrieve(redisConn, s.cfg, ilName)
+	if err != nil {
+		mlLog.WithFields(log.Fields{
+			"error":     err.Error(),
+			"component": "statestorage",
+			"key":       ilName,
+		}).Error("State storage error")
+
+		stats.Record(fnCtx, MlGrpcErrors.M(1))
+		return &mmlogic.Roster{}, err
+	}
+	// TODO: fix this
+	mlLog.Debug(fmt.Sprintf("Retreival success %v", il))
+
+	stats.Record(fnCtx, MlGrpcRequests.M(1))
+	return createRosterfromPlayerIds(il), err
+}
+
+// allIgnoreLists combines all the ignore lists and returns them.
+func (s *mmlogicAPI) allIgnoreLists(c context.Context, in *mmlogic.IlInput) (allIgnored []string, err error) {
+
+	// Get redis connection from pool
+	redisConn := s.pool.Get()
+	defer redisConn.Close()
+
+	mlLog.Info("Attempting to get and combine ignorelists")
+
+	// Loop through all ignorelists configured in the config file.
+	for il := range s.cfg.GetStringMap("ignoreLists") {
+		ilCfg := s.cfg.Sub(fmt.Sprintf("ignoreLists.%v", il))
+		thisIl, err := ignorelist.Retrieve(redisConn, ilCfg, il)
+		if err != nil {
+			panic(err)
+		}
+
+		// Join this ignorelist to the others we've retrieved
+		allIgnored = union(allIgnored, thisIl)
+	}
+
+	return allIgnored, err
+}
+
+// Set data structure functions.
+// TODO: maybe move these into an internal module if they are useful elsewhere.
+func intersection(a []string, b []string) (out []string) {
+
+	hash := make(map[string]bool)
+
+	for _, v := range a {
+		hash[v] = true
+	}
+
+	for _, v := range b {
+		if _, found := hash[v]; found {
+			out = append(out, v)
+		}
+	}
+
+	return out
+
+}
+
+func union(a []string, b []string) (out []string) {
+
+	hash := make(map[string]bool)
+
+	// collect all values from input args
+	for _, v := range a {
+		hash[v] = true
+	}
+
+	for _, v := range b {
+		hash[v] = true
+	}
+
+	// put values into string array
+	for k := range hash {
+		out = append(out, k)
+	}
+
+	return out
+
+}
+
+func difference(a []string, b []string) (out []string) {
+
+	hash := make(map[string]bool)
+	out = append([]string{}, a...)
+
+	for _, v := range b {
+		hash[v] = true
+	}
+
+	// Iterate through output, removing items found in b
+	for i := 0; i < len(out); i++ {
+		if _, found := hash[out[i]]; found {
+			// Remove this element by moving the copying the last element of the
+			// array to this index and then slicing off the last element.
+			// https://stackoverflow.com/a/37335777/3113674
+			out[i] = out[len(out)-1]
+			out = out[:len(out)-1]
+		}
+	}
+
+	return out
+}
+
+// Functions for getting or setting player IDs to/from rosters
+// Probably should get moved to an internal module in a future version.
+func getPlayerIdsFromRoster(r *mmlogic.Roster) []string {
```

Comment:
Helper inside internal?

Reply:
Probably a 'companion' golang source file for working with protobuf messages that lives in internal/pb but isn't auto-generated from protoc, I think.